R/base.R

```diff
@@ -47,25 +47,22 @@ base_params_set <- function(theme = .globals$theme) {
 
 base_params_restore <- function() {
   if (is.null(.globals$base_params)) return()
-  do.call(par_no_new_device, .globals$base_params)
+  do.call(attempt_par, .globals$base_params)
   rm("base_params", envir = .globals)
 }
 
-par_no_new_device <- function(..., fun) {
-  with_no_new_device(..., fun = par)
-}
-
-palette_no_new_device <- function(..., fun) {
-  with_no_new_device(..., fun = palette)
+attempt_par <- function(...) {
+  if (is_null_device()) {
+    attempt_with_new_device(par(...))
+  } else {
+    par(...)
+  }
 }
 
-with_no_new_device <- function(..., fun) {
-  dev_before <- dev.cur()
-  res <- fun(...)
-  dev_after <- dev.cur()
-  if (dev_before != dev_after) {
-    dev.off(dev_after)
-    dev.set(dev_before)
+attempt_palette <- function(...) {
+  if (is_null_device()) {
+    attempt_with_new_device(palette(...))
+  } else {
+    palette(...)
   }
```

Comment:
If more methods come around in the future, it might be worth doing a wrapper method to avoid duplication

```r
attempt_ <- function(code) {
  if (is_null_device()) {
    attempt_with_new_device(code)
  } else {
    force(code)
  }
}
attempt_par <- function(...) {
  attempt_(par(...))
}
```